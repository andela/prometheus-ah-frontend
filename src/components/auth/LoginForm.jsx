import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import isEmpty from 'is-empty';
import TextField from '../common/TextField';
import loginAction, { deleteErrorMessages } from '../../actions/authentication/loginAction';
import UserInputValidation from '../../vaidations/validateLogin';
import ErrorAlertNotification from '../common/ErrorAlertNotification';
import googleLogo from '../../images/google.png';
import facebookLogo from '../../images/facebook-logo.png';

/**
 * @description Login form component
 *
 * @class LoginForm
 *
 * @param {object} event
 *
 * @extends {Component}
 */
export class LoginForm extends Component {
  state = {
    username: '',
    password: '',
    errors: {}
  };

  onChange = (event) => {
    const { errors } = this.state;
    if (errors[event.target.name]) {
      const newErrors = Object.assign({}, errors);
      delete newErrors[event.target.name];
      this.setState({
        [event.target.name]: event.target.value, errors: newErrors
      });
    } else {
      this.setState({
        [event.target.name]: event.target.value
      });
    }
  }

  onSubmit = (event) => {
    const { login } = this.props;
    event.preventDefault();
    if (this.isValid()) {
      this.setState({ errors: {} });
      login(this.state);
    }
  }

  handleDelete = () => {
    const { deleteErrorMessage } = this.props;
    deleteErrorMessage();
    this.setState({
      password: '',
      password_confirmation: '',
    });
  }

  isValid = () => {
    const { errors, isValid } = UserInputValidation.loginInputValidation(this.state);

    if (!isValid) {
      this.setState({ errors, password: '' });
    }

    return isValid;
  }

  /**
   * @description Render the JSX template
   *
   * @memberof Login
   *
   * @returns {JSX} JSX representation of component
   */
  render() {
    const { error, auth } = this.props;
    const { username, password, errors } = this.state;

    if (auth) {
      return <Redirect to="/" />;
    }

    return (
      <div className="card border-0">
        <div>
          <small className="form-text login-label">Welcome to Authors Haven</small>
        </div>
        <div className="card-body">
          {
            !isEmpty(error) && (<ErrorAlertNotification
              errors={error.message}
              onClick={this.handleDelete}
            />
            )
          }
          <form onSubmit={this.onSubmit}>
            <div className="form-row">
              <div className="form-group col-md-12">
                <TextField
                  error={errors.username}
                  label="username"
                  onChange={this.onChange}
                  value={username}
                  placeholder="username"
                  field="username"
                  type="text"
                  id="username"
                />
              </div>
              <div className="form-group col-md-12">
                <TextField
                  error={errors.password}
                  label="password"
                  onChange={this.onChange}
                  value={password}
                  placeholder="password"
                  field="password"
                  type="password"
                  id="password"
                />
              </div>
            </div>
            <div className="form-row">
              <div className="form-group col-md-12">
                <button type="submit" id="submit" className="btn login-btn" onClick={this.onSubmit}>
                    SIGN IN
                </button>
                <div className="mt-4">
                  <a href="#" className="form-text text-danger">Forgot your password?</a>
                </div>
              </div>
            </div>
            <hr />
            <div>
              <small className="form-text">sign in with one of these services</small>
              <br />
            </div>
            <div className="form-row">
              <div className="form-group col-md-6">
                <button type="submit" className="btn  social-submit-btn">
                  <img className="social-submit-logo" alt="facebooklogo" src={facebookLogo} />
                  FACEBOOK
                </button>
              </div>
              <div className="form-group col-md-6">
                <button type="submit" className="btn social-submit-btn">
                  <img className="social-submit-logo" alt="facebooklogo" src={googleLogo} />
                  GOOGLE
                </button>
              </div>
            </div>
          </form>
        </div>
      </div>
    );
  }
}

LoginForm.propTypes = {
  auth: PropTypes.bool,
  error: PropTypes.shape({}),
  login: PropTypes.func.isRequired,
  deleteErrorMessage: PropTypes.func,
};

const mapStateToProps = state => ({
  auth: state.auth.isAuthenticated,
  error: state.auth.error
});

const mapDispatchToProps = dispatch => ({
  login: user => dispatch(loginAction(user)),
  deleteErrorMessage: () => dispatch(deleteErrorMessages())
});

export default connect(mapStateToProps, mapDispatchToProps)(LoginForm);
