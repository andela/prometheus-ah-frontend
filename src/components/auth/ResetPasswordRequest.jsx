import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { startResetPassword } from '../../actions/authentication/resetPassword';
import resetPasswordValidations from '../../validations/resetPasswordValidation';
import routes from '../../constants/routes';
import TextField from '../common/TextField';


/**
 * sign in an existing user
 *
 * @class ResetPasswordRequest
 * @param {event} event

 * @extends {Component}
 */
class ResetPasswordRequest extends Component {
  state = {
    email: '',
    emailValid: false,
    error: {},
  };

  onChange = (event) => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  };

  onSubmit = (event) => {
    event.preventDefault();

    const { error, hasErrors } = resetPasswordValidations.emailInputValidation(
      this.state
    );
    this.setState(() => ({
      error
    }));
    if (!hasErrors) {
      const { email } = this.state;
      const { resetPasswordRequest } = this.props;
      resetPasswordRequest(email)
        .then((payload) => {
          if (payload) {
            this.setState(() => ({
              emailValid: true
            }));
          }
        })
    }
  };

  /**
   * renders component to DOM
   *
   * @memberof PasswordReset
   *
   * @returns {JSX} JSX representation of component
   */
  render() {
    const { error, email, emailValid } = this.state;
    if (emailValid) {
      return <Redirect to={routes.LANDING} />;
    }
    return (
      <div className="form-gap">
        <div className="container">
          <div className="form-row text-center justify-content-center">
            <div className="col-md-6 col-md-offset-4 border rounded">
              <div className="panel panel-default">
                <div className="panel-body">
                  <div className="text-center">
                    <h3>
                      <i className="fa fa-lock fa-4x" />
                    </h3>
                    <h2 className="text-center">Forgot Password?</h2>
                    <p>Enter Your email address to get a password reset link</p>
                    <div className="panel-body">
                      <form
                        id="register-form"
                        autoComplete="off"
                        className="form"
                        method="post"
                        onSubmit={this.onSubmit}
                      >
                        <div className="form-group">
                          <TextField
                            error={error.email}
                            onChange={this.onChange}
                            value={email}
                            placeholder="enter your email"
                            type="email"
                            className="form-control form-input"
                            field="email"
                          />
                        </div>
                        <div className="form-group">
                          <input
                            name="recover-submit"
                            className="btn btn-lg btn-block login-btn mt-4"
                            value="Send email verification link"
                            type="submit"
                          />
                        </div>
                      </form>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

ResetPasswordRequest.defaultProps = {
  resetPasswordRequest: () => {}
};

ResetPasswordRequest.propTypes = {
  resetPasswordRequest: PropTypes.func
};

const mapDispatchToProps = dispatch => ({
  resetPasswordRequest: bindActionCreators(startResetPassword, dispatch),
});

export default connect(
  undefined,
  mapDispatchToProps
)(ResetPasswordRequest);
