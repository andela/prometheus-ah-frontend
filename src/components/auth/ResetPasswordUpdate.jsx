import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { startResetPasswordUpdate } from '../../actions/authentication/resetPassword';
import resetPasswordValidations from '../../validations/resetPasswordValidation';
import routes from '../../constants/routes';
import TextField from '../common/TextField';

/**
 * sign in an existing user
 *
 * @class ResetPasswordRequest
 * @param {event} event

 * @extends {Component}
 */
export class ResetPasswordUpdate extends Component {
  state = {
    password: '',
    password_confirmation: '',
    passwordtoken: '',
    passwordUpdated: false,
    error: {}
  };

  /**
   * @description sets the passwordtoken
   *
   * @function getDerivedStateFromProps
   *
   * @param {object} nextProps
   *
   * @returns {null} update the state
   */
  static getDerivedStateFromProps(nextProps) {
    const { location } = nextProps;
    if (location) {
      const passwordtoken = location.search.split('=')[1];

      return { passwordtoken };
    }

    return null;
  }

  onChange = (event) => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  };

  onSubmit = (event) => {
    event.preventDefault();

    const {
      error,
      hasErrors
    } = resetPasswordValidations.passwordInputValidation(this.state);
    this.setState(() => ({
      error
    }));
    if (!hasErrors) {
      const { password, password_confirmation, passwordtoken } = this.state; // eslint-disable-line
      const { resetPasswordUpdate } = this.props;
      resetPasswordUpdate({
        password,
        password_confirmation,
        passwordtoken
      })
        .then((payload) => {
          if (payload) {
            this.setState(() => ({
              passwordUpdated: true
            }));
          }
        })
    }
  };

  /**
   * renders component to DOM
   *
   * @memberof PasswordReset
   *
   * @returns {JSX} JSX representation of component
   */
  render() {
    const {
      error,
      password,
      password_confirmation, // eslint-disable-line
      passwordUpdated,
    } = this.state;

    const { location } = this.props;
    const passwordtoken = location.search.split('=')[1];

    if (!passwordtoken || passwordUpdated) {
      return <Redirect to={routes.LANDING} />;
    }

    return (
      <div className="container py-5 form-gap form-margin">
        <div className="row">
          <div className="col-md-12">
            <div className="row">
              <div className="col-md-6 mx-auto">
                <div className="card rounded-0">
                  <div className="card-header">
                    <h4 className="modal-title">Enter Password</h4>
                  </div>
                  <div className="card-body">
                    <form
                      className="form mt-2"
                      autoComplete="off"
                      id="formLogin"
                      method="post"
                      onSubmit={this.onSubmit}
                    >
                      <div className="form-group password-input">
                        <TextField
                          error={error.password}
                          onChange={this.onChange}
                          value={password}
                          placeholder="enter your password"
                          type="password"
                          className="form-control form-input"
                          field="password"
                        />
                      </div>
                      <div className="form-group password-input">
                        <TextField
                          error={error.password_confirmation}
                          onChange={this.onChange}
                          value={password_confirmation} // eslint-disable-line
                          placeholder="confirm your password"
                          type="password"
                          className="form-control form-input"
                          field="password_confirmation"
                        />
                      </div>
                      <button
                        type="submit"
                        className="btn btn-block btn-update login-btn"
                        id="btnLogin"
                      >
                        Reset Password
                      </button>
                    </form>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

ResetPasswordUpdate.defaultProps = {
  resetPasswordUpdate: () => {}
};

ResetPasswordUpdate.propTypes = {
  resetPasswordUpdate: PropTypes.func,
  location: PropTypes.shape({}).isRequired
};

const mapDispatchToProps = dispatch => ({
  resetPasswordUpdate: bindActionCreators(startResetPasswordUpdate, dispatch)
});

export default connect(
  undefined,
  mapDispatchToProps
)(ResetPasswordUpdate);
