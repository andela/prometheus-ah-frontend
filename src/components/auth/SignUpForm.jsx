import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import TextField from '../common/TextField';
import SignUpInputValidation from '../../validations/SignUpInputValidate';
import ErrorAlertNotification from '../common/ErrorAlertNotification';
import SocialLogin from '../socialLogin/SocialLogin';
import { userSignUpRequest, deleteErrorMessage } from '../../actions/signUp.action';


/**
 * @class SignUpForm
 */
export class SignUpForm extends Component {
    state = {
      username: '',
      email: '',
      password: '',
      password_confirmation: '',
      errors: {},
      success: {},
      done: false,
      isLoading: false,
      error: {},
    };


  /**
   *
   * @param {*} event
   * @returns {*} - state
   */
  onChange = (event) => {
    const { errors } = this.state;
    if (errors[event.target.name]) {
      const error = Object.assign({}, errors);
      delete errors[event.target.name];
      this.setState({
        [event.target.name]: event.target.value, error
      });
    } else {
      this.setState({
        [event.target.name]: event.target.value

      });
    }
  }

  handleDelete = () => {
    const { deleteError } = this.props;
    deleteError();
    this.setState({
      password: '',
      password_confirmation: '',
    });
  }

  /**
   *
   * @param {*} event
   * @returns {*} - state
   */
  onSubmit = (event) => {
    event.preventDefault();
    if (this.isValid()) {
      this.setState({ errors: {}, isLoading: true });
      const { signUp } = this.props;
      signUp(this.state);
    }
  }

  /**
   * isValid
   * @returns {*} - state
   */
  isValid() {
    const { errors, isValid } = SignUpInputValidation.InputValidation(this.state);
    if (!isValid) {
      this.setState({ errors });
    }
    return isValid;
  }

  /**
   *
   * @param {*} event
   * @returns {*} - state
   */
  render() {
    const { auth } = this.props;
    if (auth) {
      return <Redirect to="/users/login" />;
    }
    const {
      errors, email, username, password, password_confirmation // eslint-disable-line
    } = this.state;

    const { error } = this.props;

    const form = (
      <div className="card">
        <div>
          <small className="form-text login-label">Welcome to Authors Haven</small>
          <small className="form-text login-label-2">
          You can join the community by filling this form
          </small>
        </div>
        <div className="card-body">
          {
                (error && error.message) ? (
                  <ErrorAlertNotification
                    errors={error.message}
                    onClick={this.handleDelete}
                  />
                ) : ''
              }
          <form>
            <div className="form-row">
              <div className="form-group col-md-6">
                <TextField
                  error={errors.username}
                  onChange={this.onChange}
                  value={username}
                  placeholder="username"
                  field="username"
                  className="myusername"
                />
              </div>
              <div className="form-group col-md-6">
                <TextField
                  error={errors.email}
                  onChange={this.onChange}
                  value={email}
                  placeholder="email"
                  field="email"
                  className="myemail"
                />
              </div>
            </div>
            <div className="form-row">
              <div className="form-group col-md-6">
                <TextField
                  error={errors.password}
                  onChange={this.onChange}
                  value={password}
                  type="password"
                  placeholder="password"
                  field="password"
                  className="mypassword"
                />
              </div>
              <div className="form-group col-md-6">
                <TextField
                  error={errors.password_confirmation}
                  onChange={this.onChange}
                  value={password_confirmation} // eslint-disable-line
                  type="password"
                  placeholder="confirm password"
                  field="password_confirmation"
                  className="mypassword_confirmation"
                />
              </div>
            </div>
            <div className="form-row">
              <div className="form-group col-md-12">
                <button type="button" className="btn login-btn register" onClick={this.onSubmit}>
                      SIGN UP
                </button>
              </div>
            </div>
            <hr />
            <div>
              <small className="form-text">sign up with one of these services</small>
              <br />
            </div>
          </form>
          <SocialLogin />
        </div>
      </div>
    );
    return (
      <div>
        {form}
      </div>
    );
  }
}

SignUpForm.propTypes = {
  signUp: PropTypes.func.isRequired,
  deleteError: PropTypes.func.isRequired,
  error: PropTypes.shape({}),
  auth: PropTypes.bool
};

const mapStateToProps = state => ({
  error: state.auth.error,
  auth: state.auth.isAuthenticated
});


export default connect(mapStateToProps,
  { signUp: userSignUpRequest, deleteError: deleteErrorMessage })((SignUpForm));
