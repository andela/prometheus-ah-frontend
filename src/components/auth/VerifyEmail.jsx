import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { startVerifyEmail } from '../../actions/index';

/**
 * @class SignUpForm
 */
export class VerifyEmail extends Component {
  /**
 * @returns {*} - state
 */
  componentDidMount() {
    this.emailCheckToken();
  }

  emailCheckToken = () => {
    const { location } = this.props;
    const emailToken = location.search.split('=')[1];
    this.setState(() => ({ emailToken }));
    const { startVerifyEmailAction } = this.props;
    startVerifyEmailAction(emailToken);
  }

  /**
 * @returns {*} - state
 */
  render() {
    const { success } = this.props;
    if (success) {
      return <Redirect to="/users/login" />;
    }
    return (
      <div>
        <div>
          <h1>Verification Page</h1>
        </div>
      </div>
    );
  }
}

VerifyEmail.propTypes = {
  startVerifyEmailAction: PropTypes.func.isRequired,
  location: PropTypes.shape({}).isRequired,
  success: PropTypes.string, // eslint-disable-line
};

const mapStateToProps = state => ({
  success: state.verifyEmail.success,
});

export default connect(mapStateToProps, { startVerifyEmailAction: startVerifyEmail })(VerifyEmail);
