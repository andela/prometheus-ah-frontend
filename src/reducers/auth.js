import isEmpty from 'is-empty';
import {
  SET_CURRENT_USER,
  SIGN_UP_ERRORS,
  SIGN_UP_SUCCESS,
  DELETE_ERROR_MESSAGE,
  SET_CURRENT_USER_FAIL
} from '../actions/types';


const initialState = {
  isAuthenticated: false,
  user: {},
  error: {}
};

export default (state = initialState, action) => {
  switch (action.type) {
    case SET_CURRENT_USER:
      return {
        isAuthenticated: !isEmpty(action.user),
        user: action.user
      };

    case SET_CURRENT_USER_FAIL:
      return {
        isAuthenticated: false,
        user: {},
        error: action.error
      };

    case SIGN_UP_ERRORS:
      return {
        isAuthenticated: false,
        error: action.error
      };
    case SIGN_UP_SUCCESS:
      return {
        success: action.success
      };
    case DELETE_ERROR_MESSAGE:
      return {
        error: {}
      };
    default: return state;
  }
};
