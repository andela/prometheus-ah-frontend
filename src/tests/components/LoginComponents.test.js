import React from 'react';
import { shallow } from 'enzyme';
import { LoginForm } from '../../components/auth/LoginForm';
import mockData from '../__mocks__/mockData';

let wrapper;

describe('Login component', () => {
  beforeEach(() => {
    const props = {
      login: jest.fn(),
      auth: false,
    };
    wrapper = shallow(<LoginForm {...props} />);
  });

  it('should renders a snapshot of the app component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should set username when username changes', () => {
    const event = {
      target: {
        name: 'username',
        value: 'faksam'
      }
    };

    const username = wrapper.find('#username');

    username.simulate('change', event);

    const expectedUsername = 'faksam';

    expect(wrapper.instance().state.username).toBe(expectedUsername);
  });

  it('should set password when password changes', () => {
    const event = {
      target: {
        name: 'password',
        value: '90123456'
      }
    };

    const password = wrapper.find('#password');

    password.simulate('change', event);

    const expectedPassword = '90123456';

    expect(wrapper.instance().state.password).toBe(expectedPassword);
  });

  it('should login user when correct creadentials are supplied', () => {
    const { loginData } = mockData;
    const event = {
      preventDefault: jest.fn(),
    };

    const Login = wrapper.find('#submit');

    wrapper.setState(loginData);
    Login.simulate('click', event);
  });

  it('should return error for invalid form submission', () => {
    wrapper.find('#submit').simulate('click', {
      preventDefault: jest.fn(),
    });
    expect(wrapper.state('error')).not.toBe({});
    expect(wrapper).toMatchSnapshot();
  });
});
