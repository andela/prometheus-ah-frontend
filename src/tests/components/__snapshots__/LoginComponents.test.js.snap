// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component should renders a snapshot of the app component 1`] = `
<div
  className="card border-0"
>
  <div>
    <small
      className="form-text login-label"
    >
      Welcome to Authors Haven
    </small>
  </div>
  <div
    className="card-body"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="form-row"
      >
        <div
          className="form-group col-md-12"
        >
          <TextField
            field="username"
            id="username"
            label="username"
            onChange={[Function]}
            placeholder="username"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group col-md-12"
        >
          <TextField
            field="password"
            id="password"
            label="password"
            onChange={[Function]}
            placeholder="password"
            type="password"
            value=""
          />
        </div>
      </div>
      <div
        className="form-row"
      >
        <div
          className="form-group col-md-12"
        >
          <button
            className="btn login-btn"
            id="submit"
            onClick={[Function]}
            type="submit"
          >
            SIGN IN
          </button>
          <div
            className="mt-4"
          >
            <a
              className="form-text text-danger"
              href="#"
            >
              Forgot your password?
            </a>
          </div>
        </div>
      </div>
      <hr />
      <div>
        <small
          className="form-text"
        >
          sign in with one of these services
        </small>
        <br />
      </div>
    </form>
    <SocialLogin />
  </div>
</div>
`;

exports[`Login component should return error for invalid form submission 1`] = `
<div
  className="card border-0"
>
  <div>
    <small
      className="form-text login-label"
    >
      Welcome to Authors Haven
    </small>
  </div>
  <div
    className="card-body"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="form-row"
      >
        <div
          className="form-group col-md-12"
        >
          <TextField
            error={
              Array [
                "This username is a required field.",
              ]
            }
            field="username"
            id="username"
            label="username"
            onChange={[Function]}
            placeholder="username"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group col-md-12"
        >
          <TextField
            error={
              Array [
                "This password is a required field.",
              ]
            }
            field="password"
            id="password"
            label="password"
            onChange={[Function]}
            placeholder="password"
            type="password"
            value=""
          />
        </div>
      </div>
      <div
        className="form-row"
      >
        <div
          className="form-group col-md-12"
        >
          <button
            className="btn login-btn"
            id="submit"
            onClick={[Function]}
            type="submit"
          >
            SIGN IN
          </button>
          <div
            className="mt-4"
          >
            <a
              className="form-text text-danger"
              href="#"
            >
              Forgot your password?
            </a>
          </div>
        </div>
      </div>
      <hr />
      <div>
        <small
          className="form-text"
        >
          sign in with one of these services
        </small>
        <br />
      </div>
    </form>
    <SocialLogin />
  </div>
</div>
`;
